### ISSUE TITLE
error starting container issue in a custom kernel    

### ISSUE THREAD

**Issue by laithrafid** [docker/docker#26336]
Checkconfig.sh output : 
[root@dhcp-10-17-17-40 /]# ./check-config.sh
info: reading kernel config from root/dev/linux-kernel/.config ...

Generally Necessary:
- cgroup hierarchy: properly mounted [/sys/fs/cgroup]
- CONFIG_NAMESPACES: enabled
- CONFIG_NET_NS: enabled
- CONFIG_PID_NS: enabled
- CONFIG_IPC_NS: enabled
- CONFIG_UTS_NS: enabled
- CONFIG_DEVPTS_MULTIPLE_INSTANCES: enabled
- CONFIG_CGROUPS: enabled
- CONFIG_CGROUP_CPUACCT: enabled
- CONFIG_CGROUP_DEVICE: enabled
- CONFIG_CGROUP_FREEZER: enabled
- CONFIG_CGROUP_SCHED: enabled
- CONFIG_CPUSETS: enabled
- CONFIG_MEMCG: enabled
- CONFIG_KEYS: enabled
- CONFIG_VETH: enabled (as module)
- CONFIG_BRIDGE: enabled (as module)
- CONFIG_BRIDGE_NETFILTER: enabled (as module)
- CONFIG_NF_NAT_IPV4: enabled (as module)
- CONFIG_IP_NF_FILTER: enabled (as module)
- CONFIG_IP_NF_TARGET_MASQUERADE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_CONNTRACK: enabled (as module)
- CONFIG_NF_NAT: enabled (as module)
- CONFIG_NF_NAT_NEEDED: enabled
- CONFIG_POSIX_MQUEUE: enabled

O Features:
- CONFIG_USER_NS: enabled
  (RHEL7/CentOS7: User namespaces disabled; add 'user_namespace.enable=1' to boot command line)
- CONFIG_SECCOMP: enabled
- CONFIG_CGROUP_PIDS: missing
- CONFIG_MEMCG_SWAP: missing
- CONFIG_MEMCG_SWAP_ENABLED: missing
- CONFIG_MEMCG_KMEM: enabled
- CONFIG_RESOURCE_COUNTERS: enabled
- CONFIG_BLK_CGROUP: enabled
- CONFIG_BLK_DEV_THROTTLING: enabled
- CONFIG_IOSCHED_CFQ: enabled
- CONFIG_CFQ_GROUP_IOSCHED: enabled
- CONFIG_CGROUP_PERF: enabled
- CONFIG_CGROUP_HUGETLB: missing
- CONFIG_NET_CLS_CGROUP: enabled
- CONFIG_CGROUP_NET_PRIO: enabled
- CONFIG_CFS_BANDWIDTH: enabled
- CONFIG_FAIR_GROUP_SCHED: enabled
- CONFIG_RT_GROUP_SCHED: enabled
- CONFIG_IP_VS: enabled (as module)
- CONFIG_EXT3_FS: missing
- CONFIG_EXT3_FS_XATTR: missing
- CONFIG_EXT3_FS_POSIX_ACL: missing
- CONFIG_EXT3_FS_SECURITY: missing
    (enable these ext3 configs if you are using ext3 as backing filesystem)
- CONFIG_EXT4_FS: enabled (as module)
- CONFIG_EXT4_FS_POSIX_ACL: enabled
- CONFIG_EXT4_FS_SECURITY: enabled
- Network Drivers:
  - "overlay":
    - CONFIG_VXLAN: enabled (as module)
    Optional (for secure networks):
    - CONFIG_XFRM_ALGO: enabled
    - CONFIG_XFRM_USER: enabled
  - "ipvlan":
    - CONFIG_IPVLAN: missing
  - "macvlan":
    - CONFIG_MACVLAN: enabled (as module)
    - CONFIG_DUMMY: enabled (as module)
- Storage Drivers:
  - "aufs":
    - CONFIG_AUFS_FS: missing
  - "btrfs":
    - CONFIG_BTRFS_FS: enabled (as module)
    - CONFIG_BTRFS_FS_POSIX_ACL: enabled
  - "devicemapper":
    - CONFIG_BLK_DEV_DM: enabled (as module)
    - CONFIG_DM_THIN_PROVISIONING: enabled (as module)
  - "overlay":
    - CONFIG_OVERLAY_FS: enabled (as module)
  - "zfs":
    - /dev/zfs: missing
    - zfs command: missing
    - zpool command: missing

Limits:
- /proc/sys/kernel/keys/root_maxkeys: 1000000

**Output of `docker version`:**
docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:
 OS/Arch:      linux/amd64

**Output of `docker info`:**

docker info
Containers: 13
 Running: 0
 Paused: 0
 Stopped: 13
Images: 11
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker-253:0-315526-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.082 GB
 Data Space Total: 107.4 GB
 Data Space Available: 34.66 GB
 Metadata Space Used: 4.837 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.143 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay host bridge null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.18.3-g1d82046
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 56
Total Memory: 123.8 GiB
Name: dhcp-10-17-17-40.diablo.lan
ID: G3RO:G3UF:SBAF:JENM:PRH3:V77P:GOPC:PJ4K:6MVQ:5SRZ:QZ7P:BVET
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8

**Additional environment details (AWS, VirtualBox, physical, etc.):**

physical server  

the error i'm getting every time i start a container : 

docker: Error response from daemon: oci runtime error: process_linux.go:286: decoding sync type from init pipe caused "read parent: connection reset by peer".

**Comment by laithrafid** [comment]
3.18.3-g1d82046 #1 SMP Fri Sep 2 15:20:34 EDT 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by vieux** [comment]
@crosbymichael can you please take a look ?

**Comment by laithrafid** [comment]
thanks @vieux for the prompt response

**Comment by laithrafid** [comment]
i confirmed that the virtualization is turned on in the bios

**Comment by unclejack** [comment]
Those distributions get support only for the kernels shipped with the distribution and any updates provided through their repositories. Red Hat provides support for RHEL only with kernels they provide as packages. The CentOS community does something similar and custom kernels aren't supported.

That kernel version also seems pretty old and doesn't seem to be a mainline kernel.

This doesn't mean that the issue shouldn't be investigated. We might not be able to discover the reason for this failure because this is a custom setup (custom kernel built from modified kernel sources with an unknown kernel config).

Can this be reproduced with a non-custom kernel? What about another distribution with a different kernel?

**Comment by laithrafid** [comment]
YES i can reproduce it ... however only on this kernel ... i hadn't any problems with any other kernels !

**Comment by crosbymichael** [comment]
@laithrafid there is nothing that stands out in the output of the check config script.  So i guess congratulations, you are hitting something that we don't know about in the kernel config that can prevent containers from starting ;)

I'm not sure the best way to debug.  Maybe a diff between the kernel configs of the base that works and the changes you have in this kernel would be best to see what is different.  If you could post that it would help.

**Comment by laithrafid** [comment]
@crosbymichael Lol thanks ... looking at vimdiff of the kernel config files of the both kernels will share it as i find something interesting ...

**Comment by laithrafid** [comment]
Well what i found is that KVM configs in the kernel configs are not enabled .... and since lxc and dockers needs kvm , it must be enabled in the kernel

**Comment by crosbymichael** [comment]
@laithrafid so you enabled kvm and containers start?  Can you paste the section in the config with what you enabled?  Maybe we are doing something wrong and I want to find out where this dependency on kvm is coming from. 

Thanks for your help!

@justincormack why would kvm need to be enabled?

**Comment by laithrafid** [comment]
@crosbymichael yes ... virtualization part and kvm part in kernel config ...

**Comment by justincormack** [comment]
Are you absolutely sure this is a clean install? This does not look like a kernel issue to me.

**Comment by laithrafid** [comment]
@justincormack yes clean install because it works on all other kernels buy this one on the same system...

**Comment by justincormack** [comment]
What is the diff between the kernel configs with KVM enabled and not? Docker does not need KVM but it may need something that is a dependency of KVM too.

**Comment by laithrafid** [comment]
yes it dosen't needs kvm packages , but needs all dependencies that kvm use which lxc needs it too ... 
there's a lot of different kernel configs that i don't remember from the top of my head so i will paste it soon
